### PR TITLE
Flatten concrete definitions

### PR DESCRIPTION
**Pull request by imkiva**
in this PR, we simplified the inheritance between top-level declarations and sub-level ones introduced in #423. The approach is similar to #326 :

**TL; DR: read the code below**

- `Decl` and `Signatured` are refactored into three traits: `Telescopic`, `TopLevel` and `Resulted`
- Sub-level decls (Ctor and Field) now directly extend `Decl` (instead of `Signatured` before) like all top-level decls, the only difference is that sub-level decls do not implement the `TopLevel` trait.
- For pattern matching, `visitDecl` can now handle constructors and fields in the same `switch` clause that matches the data and struct. We don't need extra `visitCtor` and `visitField` methods anymore.
- For visiting a certain trait of one concrete definition, we can:
  - `<D extends Decl & Decl.Telescopic>` for generic functions.
  - `visitTelescopic(Decl decl, Decl.Telescopic proof)` for visitors.
- New decls should be declared in the way of:
  `[class|interface] <the-new-decl> extends CommonDecl implements <some-traits>`

See also: https://github.com/aya-prover/aya-dev/blob/4c9b212390419f234d83c51ede0e9502f7fa8e84/base/src/main/java/org/aya/concrete/stmt/Decl.java#L18-L48

### CODE REVIEW

**Review by ice1000 — APPROVED** [review]
Very good


/cc @tsao-chi

**Comment by tsao-chi** [comment]
@imkiva lgtm

**Comment by ice1000** [comment]
merge?

File: base/src/main/java/org/aya/concrete/stmt/TeleDecl.java [hunk 1]
```diff
@@ -24,24 +24,23 @@
...
  *
  * @author re-xyr
+ * @see Decl
  */
-public sealed abstract class TopTeleDecl extends BaseDecl.Telescopic implements TopLevelDecl {
-  public final @NotNull Accessibility accessibility;
+public sealed abstract class TeleDecl extends CommonDecl implements Decl.Telescopic, Decl.TopLevel, Decl.Resulted {
+  public final @NotNull Decl.Personality personality;
   public @Nullable Context ctx = null;
   public @NotNull Expr result;
```
Comment (ice1000) [inline base/src/main/java/org/aya/concrete/stmt/TeleDecl.java:35]: Maybe make sense to make them private

File: base/src/main/java/org/aya/concrete/stmt/TeleDecl.java [hunk 2]
```diff
@@ -24,24 +24,23 @@
...
-  public final @NotNull TopLevelDecl.Personality personality;
+  // will change after resolve
+  public @NotNull ImmutableSeq<Expr.Param> telescope;
+  public @Nullable Def.Signature signature;
```
Comment (ice1000) [inline base/src/main/java/org/aya/concrete/stmt/TeleDecl.java:38]: Ditto

File: base/src/main/java/org/aya/concrete/stmt/TeleDecl.java (hunk 1 above)
Comment (imkiva) [reply base/src/main/java/org/aya/concrete/stmt/TeleDecl.java:35]: Ok

File: base/src/main/java/org/aya/concrete/stmt/TeleDecl.java (hunk 2 above)
Comment (imkiva) [reply base/src/main/java/org/aya/concrete/stmt/TeleDecl.java:38]: These are used directly after pattern matching so it might be unnecessary to make them private. We can move to getters/setters in the future when a hook is needed.

**Comment by imkiva** [comment]
bors merge

File: base/src/main/java/org/aya/concrete/stmt/TeleDecl.java (hunk 2 above)
Comment (ice1000) [reply base/src/main/java/org/aya/concrete/stmt/TeleDecl.java:38]: IC, thx!
